conjur/logic/credential_provider/credential_store_factory.py

```diff
@@ -5,33 +5,62 @@
 
 This module is a factory for determining which credential store to use
 """
+# Third party
+import logging
 
 # Internals
 from conjur.constants import SUPPORTED_BACKENDS, DEFAULT_NETRC_FILE
+from conjur.interface.credentials_store_interface import CredentialsStoreInterface
+from conjur.logic.credential_provider.CredentialProviderTypes import CredentialProviderTypes
 from conjur.logic.credential_provider.file_credentials_provider import FileCredentialsProvider
 from conjur.logic.credential_provider.keystore_credentials_provider \
     import KeystoreCredentialsProvider
 from conjur.wrapper import KeystoreAdapter
+from conjur.util.Singleton import Singleton
+
 
 # pylint: disable=too-few-public-methods
-class CredentialStoreFactory:
+class CredentialStoreFactory(metaclass=Singleton):
     """
     CredentialStoreFactory
 
     This class follows the Factory pattern to determine which credential store to choose
     """
-    @classmethod
-    def create_credential_store(cls):
+
+    def __init__(self):
+        self.override_provider = None
+        self.first_time_return_insecure_store = True
+
+    def create_credential_store(self) -> CredentialsStoreInterface:
         """
         Factory method for determining which store to use
         """
+        if self.override_provider == CredentialProviderTypes.KEYSTORE:
+            return KeystoreCredentialsProvider(), f'{KeystoreAdapter.get_keyring_name()} credential store'
+        if self.override_provider == CredentialProviderTypes.NETRC:
+            self._log_netrc_warning()
+            return FileCredentialsProvider(), DEFAULT_NETRC_FILE
+
         if KeystoreAdapter.get_keyring_name() in SUPPORTED_BACKENDS:
             # TODO investigate other errors that would make a keyring not accessible
             # If the keyring is unlocked then we will use that
             if KeystoreAdapter.is_keyring_accessible():
                 # pylint: disable=line-too-long
                 return KeystoreCredentialsProvider(), f'{KeystoreAdapter.get_keyring_name()} credential store'
-            # If the keystore is not unlocked we will resort to netrc
-            return FileCredentialsProvider(), DEFAULT_NETRC_FILE
 
+        self._log_netrc_warning()
         return FileCredentialsProvider(), DEFAULT_NETRC_FILE
+
+    def override_usage(self, provider_type: CredentialProviderTypes):
+        """
+        if set to 'NETRC' create_credential_store will always return FileCredentialsProvider
+        if set to 'KEYSTORE' create_credential_store will always return KeystoreCredentialsProvider
+        """
+        self.override_provider = provider_type
+
+    def _log_netrc_warning(self):
+        if self.first_time_return_insecure_store:
```

Comment:
Really nice logic here!
A couple of things:
1. add a docstring
2. We aren't returning anything, rather logging a warning so maybe `first_time_log_insecure_store_warning`

Reply:
`check_store_on_first_run` ?